### ISSUE TITLE
-c command line option type check problem

### ISSUE THREAD

**Issue by akhmerov** [ipython/ipython#10725]
Right now (v6.1.0) the following happens:
```bash
ipython -c ".1 + .2"
...help exerpt skipped...
[TerminalIPythonApp] CRITICAL | The 'code_to_run' trait of a TerminalIPythonApp instance must be a unicode string, but a value of 0.30000000000000004 <class 'float'> was specified.
```
This does not match the behavior of python, which is able to execute the code. Expected outcome would of course be the same as if any other executable expression is passed with `-c`:
```python
Out[1]: 0.30000000000000004
```
I assume this is because of a too eager type guessing on the traitlets side.

**Comment by takluyver** [comment]
Yeah, it's a recurring problem with our command line config machinery. It tries `ast.literal_eval()` on the values to correctly parse numbers. I don't think it can easily be changed without breaking a lot of other stuff (think `--timeout 30`).

It works with an extra set of quotes:

```
ipython -c "'.1 + .2'"
```

(One set of quotes tell bash to keep it as one piece, one set tells Python it's a string)

**Comment by akhmerov** [comment]
Perhaps relaxing what `-c` expects would would work? It's either a string and if it isn't, then it was `eval`ed (and therefore is the output.

**Comment by Carreau** [comment]
> Perhaps relaxing what -c expects would work? It's either a string and if it isn't, then it was evaled (and therefore is the output.

It's complicated with how our traitlet system work. It attempt to parse everything even before it knows about the flag. Technically you can pass infinity of command flag to IPython:

```
$ ipython --Foo.bar=12.3 -c 'exit()'
```

It will work, and `12.3` will be parsed as the number 12.3 and store in the config until it –
 eventually – encounter a class `Foo` with a property `bar`. This property may be _defined_ only on the 2nd Thurdays of February of leap years, it does not matter. This is needed to allow IPython to parse config flags for extensions without loading the extensions ahead of time. 

We could try to special case `-c`, but is it worth the complexity ?

**Comment by akhmerov** [comment]
@Carreau I was simply considering making `-c` of type `traitlets.Undefined`, and do validation manually (i.e. if it is a string, `exec` it, otherwise output it).

**Comment by Carreau** [comment]
Hum, it may work. Do you want to attempt a PR ?

**Comment by akhmerov** [comment]
Sorry, not familiar with codebase, and not enough time to code dive.
